src/dataflow/render/reduce.rs

```diff
@@ -64,60 +63,118 @@ where
+            let mut columns_needed = 0;
+            for key in group_key.iter() {
+                for column in key.support() {
+                    columns_needed = std::cmp::max(columns_needed, column + 1);
+                }
+            }
+            for aggr in aggregates.iter() {
+                for column in aggr.expr.support() {
+                    columns_needed = std::cmp::max(columns_needed, column + 1);
+                }
+            }
+
+            let mut row_packer = RowPacker::new();
+            let (key_input, mut err_input): (
+                Collection<_, Result<(Row, Row), DataflowError>, _>,
+                _,
+            ) = self
+                .flat_map_ref(input, move |row| {
+                    let temp_storage = RowArena::new();
+                    let mut results = Vec::new();
+                    // First, evaluate the key selector expressions.
+                    // If any error we produce their errors as output and note
+                    // the fact that the key was not correctly produced.
+                    let datums = row.iter().take(columns_needed).collect::<Vec<_>>();
+                    for expr in group_key_owned.iter() {
+                        match expr.eval(&datums, &temp_storage) {
+                            Ok(val) => row_packer.push(val),
+                            Err(e) => {
+                                results.push(Err(e.into()));
+                            }
+                        }
+                    }
+
+                    // Second, evaluate the value selector.
+                    // If any error occurs we produce both the error as output,
+                    // but also a `Datum::Null` value to avoid causing the later
+                    // "ReduceCollation" operator to panic due to absent aggregates.
+                    if results.is_empty() {
+                        let key = row_packer.finish_and_reuse();
+                        for aggr in aggregates_owned.iter() {
+                            match aggr.expr.eval(&datums, &temp_storage) {
+                                Ok(val) => {
+                                    row_packer.push(val);
+                                }
+                                Err(e) => {
+                                    row_packer.push(Datum::Null);
+                                    results.push(Err(e.into()));
+                                }
+                            }
+                        }
+                        let row = row_packer.finish_and_reuse();
+                        results.push(Ok((key, row)));
+                    }
+                    // Return accumulated results.
+                    results
+                })
+                .unwrap();
+
+            // Demux out the potential errors from key and value selector evaluation.
+            use timely::dataflow::operators::ok_err::OkErr;
+            let (ok, err) = key_input.inner.ok_err(|(x, t, d)| match x {
```

Comment:
FYI, the comment [here](https://github.com/TimelyDataflow/timely-dataflow/blob/master/timely/src/dataflow/operators/ok_err.rs#L12) is wrong (the closure returns a `Result`, not a `bool`).

Reply:
Oh yeah that comment was totally copy/pasted (badly) from `branch.rs`. Oops. Thanks!

Reply:
https://github.com/TimelyDataflow/timely-dataflow/pull/329